{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cffi\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "import time\n",
    "import argparse\n",
    "import sys\n",
    "from src.SequenceProcessor import SequenceProcessor\n",
    "from src.FileHandler import FileHandler\n",
    "from src.pytrsomix import SeqAnalyzer,TRScalculator\n",
    "import pathlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] --input_fasta_folder_path\n",
      "                             INPUT_FASTA_FOLDER_PATH --tmin TMIN --tmax TMAX\n",
      "                             --mode MODE [--redo REDO]\n",
      "ipykernel_launcher.py: error: the following arguments are required: --input_fasta_folder_path, --tmin, --tmax, --mode\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/hubert/miniconda3/envs/TRS/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3351: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "def check_mode(value):\n",
    "    ivalue = int(value)\n",
    "    if ivalue < 0 or ivalue > 1:\n",
    "        raise argparse.ArgumentTypeError(f\"Mode must be 0 or 1, got {value}\")\n",
    "    return ivalue\n",
    "\n",
    "def check_threshold(value):\n",
    "    fvalue = float(value)\n",
    "    if fvalue < 0.8 or fvalue > 1.0:\n",
    "        raise argparse.ArgumentError(f\"Identity threshold for cdhit must be between 0.8 and 1.0, got {value}\")\n",
    "\n",
    "def main():\n",
    "    start_time = time.time()\n",
    "    parser = argparse.ArgumentParser(description='''This program extracts TRS sequences from a series of input genomes,\n",
    "                                     allows for length selection of extracted fragments, and prepares sequences for further analysis.''')\n",
    "    parser.add_argument('--input_fasta_folder_path', help=\"Path to a folder containing genomes in fasta format from which TRS sequences will be extracted\", \n",
    "                        required=True)\n",
    "    parser.add_argument('--tmin', help=\"Minimum length of TRS sequences\", required=True, type=int)\n",
    "    parser.add_argument('--tmax', help=\"Maximum length of TRS sequences\", required=True, type=int)\n",
    "    parser.add_argument('--mode', help=\"Mode of operation, must be 0 or 1\", required=True, type=check_mode)\n",
    "    parser.add_argument('--redo', help=\"Redo the analysis if results directory already exists\", required=False, type=str)\n",
    "    parser.add_argument('--cont', help=\"Continue the analysis from saved TRS results file\", required=False, type=str)\n",
    "    parser.add_argument('--email', help=\"Adress e-mail to be used for connection with NCBI databases\", required=True, type=str)\n",
    "    parser.add_argument('--threshold', help=\"Identity threshold for clustering using cdhit has to be between 0.8 and 1.0\", required=True,type=check_threshold)\n",
    "    parser.add_argument('--length_to_extract',help=\"Length of flanking sequences to be extracted from the full TRS sequence\", required=True, type=int)\n",
    "    parser.add_argument('--cd_hit_path',help=\"Path to the cd-hit-est executable\")\n",
    "    args = parser.parse_args()\n",
    "    input_fasta_folder_path = args.input_fasta_folder_path\n",
    "    tmin = args.tmin\n",
    "    tmax = args.tmax\n",
    "    mode = args.mode\n",
    "    redo = args.redo\n",
    "    cont = args.cont\n",
    "    email = args.email\n",
    "    threshold = args.threshold\n",
    "    length_to_extract = args.length_to_extract\n",
    "    cd_hit_path = args.cd_hit_path\n",
    "\n",
    "    SequenceProcessor.validate_and_set_email(email)\n",
    "\n",
    "    # Check if the provided input directory exists\n",
    "    if os.path.exists(input_fasta_folder_path):\n",
    "        print(\"Existing directory provided. Proceeding....\")\n",
    "        # Check for fasta files in the directory\n",
    "        fasta_files = [f for f in os.listdir(input_fasta_folder_path) if f.endswith('.fasta') or f.endswith('.fa')]\n",
    "        if not fasta_files:\n",
    "            print(\"No fasta files found in the provided directory. Quitting...\")\n",
    "            sys.exit(1)\n",
    "    else:\n",
    "        print(\"Provided directory path does not exist. Quitting...\")\n",
    "        sys.exit(1)\n",
    "\n",
    "    # Define the results directory and ensure it exists\n",
    "    input_fasta_folder_path_name = os.path.basename(input_fasta_folder_path)\n",
    "    results_directory = os.path.join(os.getcwd(), f\"{input_fasta_folder_path_name}_results\")\n",
    "    FileHandler.ensure_directory_exists(results_directory)\n",
    "    results_file_path = os.path.join(results_directory, \"results.txt\")\n",
    "\n",
    "    # Define the name of the CSV file that will store the results of the analysis \n",
    "    name_of_csv_file_storing_TRS_analysis_results = input_fasta_folder_path_name + \"results.csv\"\n",
    "    path_of_folder_storing_TRS_analysis_results = os.path.join(results_directory, \"TRS_output\")\n",
    "    FileHandler.ensure_directory_exists(path_of_folder_storing_TRS_analysis_results)\n",
    "    path_of_csv_file_storing_TRS_analysis_results = os.path.join(path_of_folder_storing_TRS_analysis_results, \n",
    "                                                                 name_of_csv_file_storing_TRS_analysis_results)\n",
    "\n",
    "    # Check if results already exist and redo is not specified\n",
    "    if os.path.exists(results_file_path) and not redo:\n",
    "        print(\"Results already exist and 'redo' not specified. Exiting program.\")\n",
    "        sys.exit()\n",
    "    elif os.path.exists(path_of_csv_file_storing_TRS_analysis_results) and cont and not redo:\n",
    "        combined_results = pd.read_csv(path_of_csv_file_storing_TRS_analysis_results)\n",
    "        print(\"DataFrame loaded. Resuming analysis...\")\n",
    "    else:\n",
    "        print(\"Starting analysis...\")\n",
    "\n",
    "        trs_calculators = []\n",
    "\n",
    "        # Iterate over each fasta file and calculate TRS\n",
    "        for fasta_file in fasta_files:\n",
    "            path_to_input_fasta = os.path.join(input_fasta_folder_path, fasta_file)\n",
    "            if not os.path.exists(path_to_input_fasta):\n",
    "                print(f\"File '{fasta_file}' does not exist! Skipping....\")\n",
    "                continue\n",
    "\n",
    "            # Define the TRS file path dynamically (this example assumes a static path, modify as needed)\n",
    "            trs_file = os.path.join(results_directory, 'trs.txt').encode()\n",
    "\n",
    "            try:\n",
    "                # Initialize and calculate TRS\n",
    "                trs_calculator = TRScalculator(sequence=path_to_input_fasta.encode(), trs=trs_file, tmin=tmin, tmax=tmax, mode=mode)\n",
    "                trs_calculator.calculate()\n",
    "                trs_calculators.append(trs_calculator)\n",
    "            except Exception as e:\n",
    "                print(f\"An error occurred while processing '{fasta_file}': {e}\")\n",
    "                continue\n",
    "\n",
    "        list_of_trs_results = []\n",
    "\n",
    "        # Iterate over each TRScalculator instance\n",
    "        for trs_calculator in trs_calculators:\n",
    "            # Extract results from the calculator\n",
    "            result = trs_calculator.Result\n",
    "            # Append the result to the list\n",
    "            list_of_trs_results.append(result)\n",
    "\n",
    "        # Concatenate all results into a single DataFrame\n",
    "        combined_trs_results = pd.concat(list_of_trs_results, ignore_index=True)\n",
    "\n",
    "        # Remove \">\" from >SEQ column\n",
    "        combined_trs_results['>SEQ'] = combined_trs_results['>SEQ'].str.replace(\">\", \"\")\n",
    "\n",
    "        # Save the results of the first analysis step to the CSV file\n",
    "        combined_trs_results.to_csv(path_of_csv_file_storing_TRS_analysis_results, index=False)\n",
    "        print(f\"Results saved to {path_of_csv_file_storing_TRS_analysis_results}\")\n",
    "\n",
    "    trs_time = time.time()\n",
    "    print(f\"TRS took {trs_time - start_time} seconds\")\n",
    "\n",
    "    l_chars = SequenceProcessor.adjust_input_to_range(length_to_extract,tmin)\n",
    "    r_chars = SequenceProcessor.adjust_input_to_range(length_to_extract,tmin)\n",
    "\n",
    "    combined_trs_results = SequenceProcessor.extract_sequences(combined_trs_results,l_chars,r_chars)\n",
    "\n",
    "    results_directory_after_flanks_extracted = f\"{results_directory}_L{l_chars}_R{r_chars}\"\n",
    "    results_directory_after_flanks_extracted_path = os.path.join(os.path.dirname(results_directory), results_directory_after_flanks_extracted)\n",
    "\n",
    "    print(f\"Results directory is set to: {results_directory_after_flanks_extracted_path}\")\n",
    "\n",
    "    if not os.path.exists(results_directory_after_flanks_extracted_path):\n",
    "        # Rename the existing results directory\n",
    "        os.rename(results_directory, results_directory_after_flanks_extracted_path)\n",
    "        print(f\"The results directory has been renamed to: {results_directory_after_flanks_extracted_path}\")\n",
    "    else:\n",
    "        print(f\"Directory {results_directory_after_flanks_extracted_path} already exists. Consider using a different name or removing the existing directory.\")\n",
    "\n",
    "    results_directory = results_directory_after_flanks_extracted_path\n",
    "\n",
    "    ncbi_ids = combined_trs_results[\"GENOME\"].unique().tolist()\n",
    "    organism_map = SequenceProcessor.fetch_organism_names(ncbi_ids)\n",
    "\n",
    "    #Map NCBI IDs to taxonomic names\n",
    "    combined_trs_results['Taxonomic Name'] = combined_trs_results['GENOME'].map(organism_map)\n",
    "\n",
    "    #Identify unmatched genomes\n",
    "    unmatched_genomes = combined_trs_results[combined_trs_results['Taxonomic Name'].isnull()][\"GENOME\"].unique()\n",
    "\n",
    "    if len(unmatched_genomes) > 0:\n",
    "        print(f\"Warning: Some genome IDs could not be matched with taxonomic names: {unmatched_genomes}\")\n",
    "    \n",
    "    # Extract sequences and create sequence IDs\n",
    "    combined_trs_results['L_id'] = combined_trs_results['Taxonomic_Name'] + '_L' + combined_trs_results['L-No'].astype(str)\n",
    "    combined_trs_results['R_id'] = combined_trs_results['Taxonomic_Name'] + '_R' + combined_trs_results['R-No'].astype(str)\n",
    "    sequences_df = combined_results[['SEQ_L', 'SEQ_R', 'L_id', 'R_id']]\n",
    "    \n",
    "    path_of_folder_storing_TRS_analysis_results = os.path.join(results_directory, \"TRS_output\")\n",
    "    fasta_files_with_flanks = os.path.join(path_of_folder_storing_TRS_analysis_results,\"combined_sequences.fasta\")\n",
    "    with open(fasta_files_with_flanks, 'w') as fasta_file:\n",
    "        for _, row in sequences_df.iterrows():\n",
    "            # Write left sequence\n",
    "            fasta_file.write(f'>{row[\"L_id\"]}\\n')\n",
    "            fasta_file.write(f'{row[\"SEQ_L\"]}\\n')\n",
    "            # Write right sequence\n",
    "            fasta_file.write(f'>{row[\"R_id\"]}\\n')\n",
    "            fasta_file.write(f'{row[\"SEQ_R\"]}\\n')\n",
    "    fasta_files_with_flanks_unique = os.path.join(path_of_folder_storing_TRS_analysis_results,\"combined_sequences_unique.fasta\")\n",
    "    SequenceProcessor.rename_sequences(fasta_files_with_flanks,fasta_files_with_flanks_unique)\n",
    "\n",
    "    cd_hit_results_folder = os.path.join(results_directory,\"cd-hit-results\")\n",
    "    FileHandler.ensure_directory_exists(cd_hit_results_folder)\n",
    "    cd_hit_output_file = os.path.join(cd_hit_results_folder,\"combined_sequences_unique_cdhit\")\n",
    "    if cd_hit_path:\n",
    "        cd_hit_path = cd_hit_path\n",
    "    elif not cd_hit_path:\n",
    "        cdhit_path = FileHandler.find_file_by_name(file_name='cd-hit-est', auto=True)\n",
    "        cd_hit_path = pathlib.Path(cd_hit_path)\n",
    "        cdhit_path = cdhit_path.parent\n",
    "    results_directory = SequenceProcessor.run_cdhit(cdhit_path, input_file=fasta_files_with_flanks_unique, output_file=cd_hit_output_file,\n",
    "                                           results_directory=results_directory, sc = 1, c = threshold)\n",
    "    \n",
    "    cdhit_clusters_file = os.path.join(cd_hit_results_folder,\"combined_sequences_unique_cdhit.clstr\")\n",
    "    non_unique_sequences = os.path.join(cd_hit_results_folder,\"combined_sequences_clusters.txt\")\n",
    "    SequenceProcessor.extract_sequences(cdhit_clusters_file,non_unique_sequences)\n",
    "    clusters_to_be_cleaned = non_unique_sequences\n",
    "    SequenceProcessor.clean_sequence_ids(clusters_to_be_cleaned)\n",
    "    clusters_cleaned = clusters_to_be_cleaned\n",
    "\n",
    "    fasta_ids_to_remove_because_they_were_in_clusters = FileHandler.read_fasta_ids(clusters_cleaned)\n",
    "    sequences_after_clusters_filtering_folder = os.path.join(results_directory,\"filtered_sequences\")\n",
    "    FileHandler.ensure_directory_exists(sequences_after_clusters_filtering_folder)\n",
    "\n",
    "    fasta_with_clustered_ids_removed = os.path.join(sequences_after_clusters_filtering_folder,\"not_in_clusters_combined_sequences_unique.fasta\")\n",
    "    fasta_with_clustered_ids_included = os.path.join(sequences_after_clusters_filtering_folder,\"in_clusters_combined_sequences_unique.fasta\")\n",
    "\n",
    "    FileHandler.filter_fasta_file(fasta_files_with_flanks_unique,fasta_with_clustered_ids_removed,fasta_ids_to_remove_because_they_were_in_clusters)\n",
    "    FileHandler.filter_fasta_file_clusters(fasta_files_with_flanks_unique,fasta_with_clustered_ids_included,fasta_ids_to_remove_because_they_were_in_clusters)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TRS",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
